fixed the bugs found during testing. Added a dust animation to fireball

0shine0 opened the pull request:
Pinging @Cervator @xrtariq2594 @portokaliu @niksoc for review.

niksoc commented on src/main/java/org/terasology/combatSystem/weaponFeatures/systems/collisionHandling/ExplosionHandlingSystem.java:
```diff
@@ -113,13 +141,6 @@ private void doExplosion(EntityRef entity){
                     entity.send(new HurtEvent(otherEntity));
```
this if condition and the previous if condition check for the same thing. the previous one is redundant and can be removed

0shine0 replied:
I don't get what you meant by this either. I don't see any 2 if conditions that check the same thing.

niksoc commented on src/main/java/org/terasology/combatSystem/weaponFeatures/systems/collisionHandling/BouncingHandlingSystem.java:
```diff
@@ -43,21 +43,28 @@ private void bounce(EntityRef entity, EntityRef target, Vector3f normal){
         
-        if(checkPeircing(normal, mass.velocity, bounce.maxPierceAngle, bounce.minVelocity)){
+        // check peircing
+        if(checkPeircing(normal, mass.velocity, bounce.maxPierceAngle, bounce.minPeirceVelocity)){
```
pierce spelling is incorrect, it's an easy fix in an ide so please fix it.

0shine0 replied:
corrected

niksoc commented on src/main/java/org/terasology/combatSystem/weaponFeatures/systems/collisionHandling/ExplodeHandlingSystem.java:
```diff
@@ -22,7 +24,17 @@
-    public void explosionOnContact(CollideEvent event, EntityRef entity){
+    public void explodeOnContact(CollideEvent event, EntityRef entity){
+        // damage the other entity
+        HurtingComponent hurting = entity.getComponent(HurtingComponent.class);
```
why is the hurting component everywhere? didn't you move it into it's own system?

0shine0 replied:
corrected

0shine0 replied:
corrected

niksoc commented on src/main/java/org/terasology/combatSystem/weaponFeatures/systems/collisionHandling/BouncingHandlingSystem.java:
```diff
@@ -43,21 +43,28 @@ private void bounce(EntityRef entity, EntityRef target, Vector3f normal){
+            entity.saveComponent(mass);
+        }
         
         entity.send(new AddCollisionExceptionEvent(target));
```
why is this here? doesn't it mean that the projectile can never bounce twice off the same object? Like if it hits one block, hits something else and then comes back, it's not impossible if one entity is huge

0shine0 replied:
corrected

0shine0 replied:
corrected

0shine0 commented on src/main/java/org/terasology/combatSystem/physics/systems/CombatPhysicsSystem.java:
```diff
@@ -121,5 +173,37 @@ public void update(float delta){
+        while(iter.hasNext()){
+            CollisionGroup collisionGroup = iter.next();
+            group[i] = collisionGroup;
+            i++;
```
corrected

0shine0 commented on src/main/java/org/terasology/combatSystem/physics/systems/CombatPhysicsSystem.java:
```diff
@@ -121,5 +173,37 @@ public void update(float delta){
+    private short combineGroups(List<CollisionGroup> collisionGroup){
+        short flag = 0;
+        Iterator<CollisionGroup> iter = collisionGroup.iterator();
+        while(iter.hasNext()){
```
corrected

0shine0 commented on src/main/java/org/terasology/combatSystem/weaponFeatures/systems/collisionHandling/BouncingHandlingSystem.java:
```diff
@@ -43,21 +43,28 @@ private void bounce(EntityRef entity, EntityRef target, Vector3f normal){
             return;
         }
         
-        if(checkPeircing(normal, mass.velocity, bounce.maxPierceAngle, bounce.minVelocity)){
```
corrected

0shine0 commented on src/main/java/org/terasology/combatSystem/weaponFeatures/systems/collisionHandling/ExplodeHandlingSystem.java:
```diff
@@ -43,34 +60,27 @@ private void explode(EntityRef entity){
+        if(explosion != null){
+            EntityRef explosionEntity = explosion.build();
+            
+            if(explosionEntity.hasComponent(ExplosionComponent.class)){
```
I don't get what you mean by that. Ideally i want the entity to explode even if the prefab does not contain explosion component. Like in case of fireball when we just explode the fireball into some dust for animation.

niksoc replied:
If you want entities to explode without the explosion component that sounds really fishy. The explosion component is what says (or atleast what should say) that an explosion will happen. an explosion damages entities in a radius, if you want a different kind of explosion then then fireball should be dealt with separately.

0shine0 replied:
The ExplodeComponent and ExplosionComponent are different here. The ExplodeComponent contains the prefab of an entity that the current entity will explode into. If the new entity has an ExplosionComponent than it will undergo a damaging explosion. ExplosionComponent contains data like the radius of explosion and the CollisionGroups that it should detect and try to damage.

The ExplodeComponent can also enable an entity to explode into another and not relating to explosion like a hat exploding into rabbit (terrible example).

I think the implementation is correct in that sense.

niksoc replied:
What I don't get is why do you check if explosion component is present in the entity or not before you send the event. The event handlers for the entities having explosion component and for the hat exploding into a rabbit will be different right? The first one should simply make explosion component compulsory.
In a different sense I thought different explosion prefabs were to have different shapes, radius etc. In that case maybe explode event should be something like transform or replace event? Generally explode does not mean suddenly changing entity.

0shine0 replied:
The first one does make exlosion component necessary. The second one is already handled in the above method. I send a ExlosionEvent because if an entity explodes into another entity that has ExplosionComponent than the other entity should undergo explosion. Meanwhile there can also be an entity that has explosion component from start. For eg. A barrel full of gun powder. If that entity will be destroyed than an explosion will occur for that entity. Also if a ExplosionEntity is sent to the entity than that entity will undergo explosion. For eg. Player throws a c4. It lies in the world waiting. Player presses the switch which send ExplosionEvent to c4 to trigger explosion. Now the above method sends ExplosionEvent. It can be sent without checking for ExplosionComponent but it reduces computation power to check for ExplosionComponent beforehand, so that it won't search through every event handler and then do nothing because ExplosionComponent was not present in the entity.

0shine0 commented:
@niksoc i corrected most of the mistakes but i don't get the last two. Also added critChance in MeleeComponent for a variable critChance for every weapon.

niksoc commented on src/main/java/org/terasology/combatSystem/weaponFeatures/systems/collisionHandling/ExplosionHandlingSystem.java:
```diff
@@ -124,22 +120,16 @@ private void doExplosion(EntityRef entity){
+            float distanceSq = location.getWorldPosition().distanceSquared(otherEntityLocation.getWorldPosition());
+            float radiusSquared = explosion.radius*explosion.radius;
+            float explosionFactor = 1 - (distanceSq/radiusSquared);
+            if(explosionFactor < 0){
```
explosion factor is < 0 when distanceSq > radiusSquared (from the formula). there's no use in setting explosion factor =  0 as the next if won't be true anyway

0shine0 replied:
Its to take care of any floating point errors during the above calculations. If the explosion factor becomes like -0.00000001 that it turns that into 0.

niksoc replied:
- You set explosionfactor to 0 when explosionfactor<0.
- explosionfactor <0 means that distanceSq/radiusSq > 1
- implying that distanceSq > radiusSq.
- means that the next if condition which requires the opposite ( distanceSq <= radiusSq) won't execute.
- means that if you set explosionFactor to 0, it will never be used anyway. 
- therefore the if condition is unnecessary.

0shine0 replied:
Oh yeah, i just thought that division may bring some small errors. I never thought it in that sense. I will correct it :)

niksoc commented on src/main/java/org/terasology/combatSystem/weaponFeatures/systems/MeleeAttackSystem.java:
```diff
@@ -18,32 +18,32 @@
-                entity.send(new CritHurtEvent(otherEntity));
-            }
+            Random rand = new Random();
+            int amount = rand.nextInt(melee.maxDamage - melee.minDamage + 1) + melee.minDamage;
```
for hurting handling system I had something different in mind. I mean to decouple hurting itself from the rest of the physics system. So CritHurtEvent should not be mentioned here. For example if sword or an arrow "hits" something, it should simply send a "hit" event. The sword itself does not handle the logic for the damage it can cause. All that should be done in HurtingSystem. In hurting system handle the "hit" event.  I also see that melee and crit are bundled together (which need not be crit damage can be caused by anything). Separate crit into its own component. So to summarize: send hit event on collision, the hurting system intercepts the event and decides what kind/how much damage it should deal based on the components (crit hurt, normal hurt whatever).

0shine0 replied:
It does make sense. I made the changes and i will upload them in the next pr.

Cervator commented:
Tested and merged!

niksoc commented on src/main/java/org/terasology/combatSystem/weaponFeatures/systems/HurtingHandlingSystem.java:
```diff
@@ -23,6 +23,11 @@ public void hurting(HurtEvent event, EntityRef entity){
         }
         
+        // checking if the damage can crit. If yes than letting the critHurting function to handle it.
+        if(entity.hasComponent(CritDamageComponent.class)){
```
instead of this, make this handler low priority (think that default behavior -> low priority) and consume the event in crit damage handler. In general you should write code that is " open to extension and closed to modification ".  If a modder wants to add a new damage type, then he has to modify this if condition to include his damage type as well.
